[JSC] Implement HasVarDeclaration abstract operation

shvaikalesh opened the pull request:
#### 66468f0cb22104024c0485895b647744b943eaab
<pre>
[JSC] Implement HasVarDeclaration abstract operation
<a href="https://bugs.webkit.org/show_bug.cgi?id=261179">https://bugs.webkit.org/show_bug.cgi?id=261179</a>
&lt;rdar://problem/115014147&gt;

Reviewed by Yusuke Suzuki.

Prior to this change, during script evaluation, variable bindings created via eval() -- including
those for hoisted block-level function declarations (Annex B) -- were ignored when checking for
duplicates with lexical declarations.

HasVarDeclaration [1] is introduced to remedy that. It correctly handling static global properties
(like NaN) that only pose as variables for performance reasons, which is achieved due to the fact that
static global properties are declared in the very first place, by the finishCreation() overrides,
and consecutively. So instead of maintaing a list of their identifiers, or adding a SymbolTableEntry
flag, we only need to keep an offset of the last such property.

Also, this change reorders duplicate variable checks into a single loop to match the spec [2],
making error messages more precise, and adds an optimization to avoid looking up each lexical
declaration if JSGlobalLexicalEnvironment is empty.

Even prior to this change, JSGlobalObject did override deleteProperty(), so no possible regression
is introduced.

Aligns JSC with the spec and SpiderMonkey yet not with V8 yet.

[1] <a href="https://tc39.es/ecma262/#sec-hasvardeclaration">https://tc39.es/ecma262/#sec-hasvardeclaration</a>
[2] <a href="https://tc39.es/ecma262/#sec-globaldeclarationinstantiation">https://tc39.es/ecma262/#sec-globaldeclarationinstantiation</a> (step 3)

* JSTests/stress/global-add-function-should-not-be-shadowed-by-lexical-bindings.js:
* JSTests/stress/global-add-var-should-not-be-shadowed-by-lexical-bindings.js: Added.
* JSTests/stress/has-var-declaration.js: Added.
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::createErrorForDuplicateGlobalVariableDeclaration):
(JSC::createErrorForInvalidGlobalAssignment): Deleted.
* Source/JavaScriptCore/runtime/ExceptionHelpers.h: Removed unused createErrorForInvalidGlobalAssignment().
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::deleteProperty):
(JSC::JSGlobalObject::createGlobalFunctionBinding):
(JSC::JSGlobalObject::addStaticGlobals):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::hasVarDeclaration):
(JSC::JSGlobalObject::createGlobalVarBinding):
* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::ProgramExecutable::initializeGlobalProperties):

Canonical link: <a href="https://commits.webkit.org/267891@main">https://commits.webkit.org/267891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cadec7296d2a303958969d9d95c918e6e497fa3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/17980 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/18307 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/18872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/19811 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/16828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/21602 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/18465 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/19811 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/18196 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/21602 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/18872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/20684 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/21602 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/18872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/20684 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/15571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/21602 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/18872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/20684 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/17255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/17129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/18465 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/21328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/16229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/18872 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/21328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/20589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/24/builds/22561 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2206 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/16978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/24/builds/22561 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -539,6 +539,8 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     bool m_needsSiteSpecificQuirks { false };
     unsigned m_globalLexicalBindingEpoch { 1 };
+    ScopeOffset m_lastStaticGlobalOffset;
+    HashSet<UniquedStringImpl*> m_varNamesDeclaredViaEval;
```
Should we use IdentifierSet?

shvaikalesh replied:
Thanks Yusuke, switched to `IdentifierSet`.

Constellation reviewed (approved):
r=me with comment.

Commit:
[JSC] Implement HasVarDeclaration abstract operation
https://bugs.webkit.org/show_bug.cgi?id=261179
<rdar://problem/115014147>

Reviewed by Yusuke Suzuki.

Prior to this change, during script evaluation, variable bindings created via eval() -- including
those for hoisted block-level function declarations (Annex B) -- were ignored when checking for
duplicates with lexical declarations.

HasVarDeclaration [1] is introduced to remedy that. It correctly handling static global properties
(like NaN) that only pose as variables for performance reasons, which is achieved due to the fact that
static global properties are declared in the very first place, by the finishCreation() overrides,
and consecutively. So instead of maintaing a list of their identifiers, or adding a SymbolTableEntry
flag, we only need to keep an offset of the last such property.

Also, this change reorders duplicate variable checks into a single loop to match the spec [2],
making error messages more precise, and adds an optimization to avoid looking up each lexical
declaration if JSGlobalLexicalEnvironment is empty.

Even prior to this change, JSGlobalObject did override deleteProperty(), so no possible regression
is introduced.

Aligns JSC with the spec and SpiderMonkey yet not with V8 yet.

[1] https://tc39.es/ecma262/#sec-hasvardeclaration
[2] https://tc39.es/ecma262/#sec-globaldeclarationinstantiation (step 3)

* JSTests/stress/global-add-function-should-not-be-shadowed-by-lexical-bindings.js:
* JSTests/stress/global-add-var-should-not-be-shadowed-by-lexical-bindings.js: Added.
* JSTests/stress/has-var-declaration.js: Added.
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::createErrorForDuplicateGlobalVariableDeclaration):
(JSC::createErrorForInvalidGlobalAssignment): Deleted.
* Source/JavaScriptCore/runtime/ExceptionHelpers.h: Removed unused createErrorForInvalidGlobalAssignment().
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::deleteProperty):
(JSC::JSGlobalObject::createGlobalFunctionBinding):
(JSC::JSGlobalObject::addStaticGlobals):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::hasVarDeclaration):
(JSC::JSGlobalObject::createGlobalVarBinding):
* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::ProgramExecutable::initializeGlobalProperties):

Canonical link: https://commits.webkit.org/267891@main
